### PR TITLE
remove duplicate linden client with same host and port

### PR DESCRIPTION
**Pull request by intergret**
linden client may announce to zk again when zk recovery. check linden client, remove duplicate ones, only keep the newest one.

### CODE REVIEW

File: linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java [hunk 1]
```diff
@@ -40,73 +45,93 @@
...
+              lindenClients.remove(node);
+              LOGGER.warn("Linden node {} {} is duplicated in shard {}, removed!", node, hostPort, shardId);
+            } else {
+              uniqueClients.add(hostPort);
+            }
+          }
+        }
+
+        LOGGER.info("{} Linden node in shard {}.", lindenClients.size(), shardId);
+        clients = new ArrayList<>();
```
Comment (yozhao) [inline linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java]: Don't assign new value to clients in a longer process. Use a temporary variable first, then assign to clients.

File: linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java [hunk 2]
```diff
@@ -40,73 +45,93 @@
...
+        // ensure the new node overrides the old node.
+        List<String> sortedNodes = new ArrayList<>();
+        for (String node : lindenClients.keySet()) {
+          sortedNodes.add(node);
+        }
+        Collections.sort(sortedNodes, Collections.reverseOrder());
+
+        Set<String> uniqueClients = new HashSet<>();
+        for (String node : sortedNodes) {
+          if (lindenClients.containsKey(node)) {
```
Comment (yozhao) [inline linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java]: no need check

Comment (intergret) [inline linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java]: the check is removed.

File: linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java (hunk 1 above)
Comment (intergret) [inline linden-core/src/main/java/com/xiaomi/linden/cluster/ShardClient.java]: Done. Also volatile the clients field, make sure other threads will see and use the new reference as discuss here: http://stackoverflow.com/questions/1351223/thread-safe-setting-of-a-variable-java

**Pull request merged**